### PR TITLE
Rewrite redirects to functionality to execute after a server-side 404 …

### PR DESCRIPTION
**Pull request by jonsherrard**
…only. No handlers set any more (they\'re expensive)

#### What have you done
I've rewritten how redirects work. They no longer set a computationally expensive handler per redirect using Hapi.

Instead, after the handler has run, we check if Hapi is _about_ to respond with a 404. Then it's last chance saloon, and we check if the origin request path is in our redirects object (in-memory), if we find it, then we hijack the reply and send our own redirect instead of the old 404

#### Why have you done it
It was taking 30 minutes to set 30k handlers

#### Testing carried out to prevent breaking changes
All tests passing, and the `redirects.json` handlers, both file and remote endpoint tests, now use a 50k line redirects file.

I've also tried this on a large project with 25k redirects, and it boots up instantly.

### Making useless requests:

The only problem is we fire off and cache a lot of HTML and API responses, mostly the 404 page, so we could potentially do this *before* any routing/requesting/rendering happens using the `onPreResponse` handler instead of `onPostHandler`

### Doesn't fix client side nav redirects

We could do something like like ask the 404 page component to hard refresh itself when you client side nav to it (`componentDidMount`). That'd then check the server for a redirect... Maybe leave that to user land.

### Something else... Memory

Something else.... the first thing I came here to write... but have now forgotten.

Ah yes! What are the ramifications of loading large redirects files into Memory? 50k redirects = about 4.5 mb of JSON.

### CODE REVIEW

File: src/server/handle-redirects.js [hunk 1]
```diff
@@ -4,34 +4,35 @@ import fetch from 'isomorphic-fetch'
...
-            .rewritable(true)
-        }
-      })
-    })
+  server.ext('onPostHandler', (request, reply) => {
+    const status = reply.request.response.statusCode
+    if (
+      (status == 404)
+      && redirects
+      && (request.url.pathname in redirects)
```
Comment (jonsherrard) [inline src/server/handle-redirects.js]: How performant is this? Are there faster methods?

File: src/server/handle-redirects.js [hunk 2]
```diff
@@ -4,34 +4,35 @@ import fetch from 'isomorphic-fetch'
...
-            .redirect(`${redirects[fromPath]}${request.url.search}`)
-            .permanent()
-            .rewritable(true)
-        }
-      })
-    })
+  server.ext('onPostHandler', (request, reply) => {
+    const status = reply.request.response.statusCode
+    if (
+      (status == 404)
```
Comment (mathewhawley) [inline src/server/handle-redirects.js]: is the `==` safe?

**Comment by thomasdigby** [comment]
This is looking real good, in reference to the _bad_ things:

- Useless requests should be ok once we clear out the backlog of old requests over the next couple of weeks, users should be sharing and navigating to the new URLs once it's bedded in a bit.
- It's unlikely we'll need a client side re-direct aside from custom links added through the CMS (within a button or paragraph widget). Could we handle it within AsyncProps `loadProps` or something similar? if 404 check the redirects file...? Pfff, I dunno. I guess refreshing the page is a gooden for now.
- I think we've got a heck lotta memory to use up so 4.5mb _should_ be fine?

**Comment by jonsherrard** [comment]
@thomasdigby 

- Yes - agreed

- Potentially - but I don't think we have an API to query the redirects file... we'd have to expose the CMS endpoint?

- Yeah I think it'll be fine - let's just get it running and then look at Heroku logs

File: src/server/handle-redirects.js (hunk 1 above)
Comment (mathewhawley) [reply src/server/handle-redirects.js]: What's the `in` syntax? Never seen it before – is it a new JS feature?

Comment (jonsherrard) [reply src/server/handle-redirects.js]: It's certainly ES5, maybe even ES3!

Comment (jonsherrard) [reply src/server/handle-redirects.js]: `key in object` basically

Comment (mathewhawley) [reply src/server/handle-redirects.js]: wow – learned something new - before my time!

Comment (thomasdigby) [reply src/server/handle-redirects.js]: It's probably negligible, but `typeof redirects[request.url.pathname] !== undefined` seems quicker in this benchmark https://jsperf.com/array-indexof-vs-object-s-in-operator/4

**Comment by jonsherrard** [comment]
@thomasdigby @mathewhawley Strict equality, and faster object check are in. LGTM?

**Pull request merged**